tests/test_app.py

```diff
@@ -0,0 +1,242 @@
+
+
+def test_show_url_none(client: FlaskClient, test_database: None) -> None:
+    # given
+    url_id = 0
+    route = f"/urls/{url_id}"
+
+    # when
+    response = client.get(route)
+
+    # then
+    expected_html = resolve_path("fixtures/expected_page_not_found.html")
+    assert response.data == expected_html.read_bytes()
+    assert response.status_code == 200
+
+
+@freeze_time("2024-05-01")
+def test_list_urls(client: FlaskClient, test_database: None) -> None:
+    # given
+    url_name = "https://www.example.com"
+    url_id = create_url(url_name)
+    url_data = get_url_data(url_id)
+    route = "/urls"
+
+    with patch(
+        "page_analyzer.utils.db.check_url",
+        return_value={
+            "response_code": "200",
+            "h1": "",
+            "title": "",
+            "description": "",
+        },
+    ):
+        create_check(url_data)
+
+    # when
+    response = client.get(route)
+
+    # then
+    expected_html = resolve_path("fixtures/expected_urls.html")
+    assert response.data == expected_html.read_bytes()
+    assert response.status_code == 200
+
+
+def test_list_urls_empty(client: FlaskClient, test_database: None) -> None:
+    # given (no urls in DB)
+    route = "/urls"
+
+    # when
+    response = client.get(route)
+
+    # then
+    expected_html = resolve_path("fixtures/expected_urls_empty.html")
+    assert response.data == expected_html.read_bytes()
+    assert response.status_code == 200
+
+
+@freeze_time("2024-05-01")
+def test_add_url_check(client: FlaskClient, test_database: None) -> None:
+    # given
+    url_name = "https://www.example1.com"
+    url_id = create_url(url_name)
+    route = f"/urls/{url_id}/checks"
+
+    # when
+    with patch(
+        "page_analyzer.utils.db.check_url",
+        return_value={
+            "response_code": "200",
+            "h1": "some header",
+            "title": "some title",
+            "description": "some description",
+        },
+    ):
+        response = client.post(route, follow_redirects=True)
+
+    # then
+    expected_html = resolve_path("fixtures/expected_url_checks_add.html")
+    assert response.data == expected_html.read_bytes()
+    assert response.status_code == 200
+    url_found_checks_data = get_url_checks(url_id)
+    assert url_found_checks_data == [
+        {
+            "id": 1,
+            "url_id": 1,
+            "status_code": 200,
+            "h1": "some header",
+            "title": "some title",
+            "description": "some description",
+            "created_at": datetime.date(2024, 5, 1),
+        }
+    ]
+
+
+@freeze_time("2024-05-01")
+def test_add_url_check_error(client: FlaskClient, test_database: None) -> None:
+    # given
+    url_name = "https://www.example1.com"
+    url_id = create_url(url_name)
+    route = f"/urls/{url_id}/checks"
+
+    # when
+    with patch("page_analyzer.utils.db.check_url", return_value=None):
+        response = client.post(route, follow_redirects=True)
```

Comment:
should this part raise instead? test function says `error`

Reply:
I consider this side effect could be raised inside check_url test.
we get None here as a result of check_url raise error.